manifold/src/impl.cpp

```diff
@@ -362,13 +362,37 @@ Manifold::Impl::Impl(Shape shape) {
  * ID can be found using the meshID2Original mapping.
  */
 void Manifold::Impl::DuplicateMeshIDs() {
-  std::map<int, int> old2new;
-  for (BaryRef& ref : meshRelation_.triBary) {
-    if (old2new.find(ref.meshID) == old2new.end()) {
-      old2new[ref.meshID] = meshID2Original_.size();
-      meshID2Original_.push_back(meshID2Original_[ref.meshID]);
-    }
-    ref.meshID = old2new[ref.meshID];
+  std::set<int> newMeshID;
+  int first;
+  {
+    auto lock = std::scoped_lock(meshID2OriginalMutex_);
+    first = meshID2Original_.size();
+    meshID2Original_.insert(meshID2Original_.end(),
+                            meshRelation_.allMeshID.begin(),
+                            meshRelation_.allMeshID.end());
+  }
+  newMeshID.insert(
+          countAt(first),
+          countAt(first + (int)meshRelation_.allMeshID.size()));
+  VecDH<int> oldIDs;
+  for (int oldID : meshRelation_.allMeshID) {
+    oldIDs.push_back(oldID);
+  }
+  meshRelation_.allMeshID = std::move(newMeshID);
+
+  auto begin = oldIDs.begin();
+  auto end = oldIDs.end();
+  VecDH<char> seen(oldIDs.size(), 0);
+  char* seenD = seen.ptrD();
+  thrust::for_each(thrust::device, meshRelation_.triBary.begin(), meshRelation_.triBary.end(),
+    [=](BaryRef& ref) {
+      int index = thrust::lower_bound(begin, end, ref.meshID) - begin;
+      seenD[index] = 1;
+      ref.meshID = first + index;
+    });
+  for (int i = 0; i < oldIDs.size(); i++) {
+    if (!seen[i])
+      meshRelation_.allMeshID.erase(first + i);
```

Comment:
This function has gotten quite a bit more complicated; maybe a comment about what we're doing here? I'm having some trouble following.